x-pack/platform/plugins/shared/onechat/server/services/tools/esql/utils/execute_esql_query.ts

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { EsqlTool } from '@kbn/onechat-server';
+import { z } from '@kbn/zod';
+import { EsqlToolCreateResponse } from '../../../../../common/tools';
+
+export const esqlSchema = z.object({
+  params: z
+    .record(z.string())
+    .optional()
+    .describe('Params needed to execute the query - must match defined tool parameter keys'),
+});
+
+export const esqlToolCreator = (tool: EsqlToolCreateResponse): EsqlTool => {
+  const executableTool: EsqlTool = {
+    id: tool.id,
+    name: tool.name,
+    description: tool.description,
+    query: tool.query,
+    params: tool.params,
+    schema: esqlSchema,
+    handler: async ({ params }, { esClient }) => {
+      const client = esClient.asCurrentUser;
+      const filledQuery = tool.query.replace(/\?(\w+)/g, (_, key) => {
```

Comment:
Should this logic (mostly regex operation now) be a proper function with unit tests? We really need to be able to detect regressions to logic here.

I'm just very bad at reading regex, so test suite would help understand what's going on in here